doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,92 @@
+# KDB operation sequences
+
+## Problem
+
+The current public API allows various kind of sequences of `kdbGet` and `kdbSet` calls.
+There is no real guideline of allowed and/or prohibited sequences, other than `kdbGet` for a certain parent key must be called before `kdbSet`:
+
+- initially after `kdbOpen`
+- after conflict errors in `kdbSet`
+
+While researching ways to do change tracking, we discovered that some sequences are currently legal, but might be problematic.
+This issue not only effects change tracking, but may cause problems for other use cases too.
+
+```c
+Key * keyA = ...;
+Key * keyB = ...;
+// assume that neither A below B nor B below A
+KeySet * a = ...;
+KeySet * b = ...;
+kdbGet(kdb, a, keyA); // (A)
+kdbGet(kdb, b, keyB); // (B)
+kdbSet(kdb, a, keyA); // (C)
+```
+
+- At point `(A)` the plugins internally store the KeySet for key `a`
+- At point `(B)` the plugins internally override the stored the KeySet with the KeySet for key `b`
+- At point `(C)` they will generate a wrong changeset for key `a`
+
+## Constraints
+
+- As far as possible, we must check for illegal sequences and raise an error when a constraint is violated
+
+## Assumptions
+
+1.
+2.
+3.
+
+## Considered Alternatives
+
+1. Document which sequences are allowed without raising errors in all illegal cases.
+   Least-effort approach, but at least offers some transparency for developers.
+   Developers utilizing Elektra can still ignore it, and may cause problems in certain setups without knowing.
+
+2. Enforce that the `parentKey` used in `kdbSet` is below the one used in the last `kdbGet`.
+   If this is not the case, `kdbSet` will abort and report an error in `parentKey`.
+   Developers might still wrongly mix the sequences, but they will get an error and have to fix it.
+
+   - Adding a `Key * lastGetParent` to `struct _KDB`
```

Comment:
Don't like this.

Reply:
I don't like that we get too much state/too many illegal sequences. To e.g. force the same parent key or keyset breaks the whole idea of separating these 3 classes.

Reply:
It is not too bad. If you always do a `kdbGet`, then modify the keyset and then do `kdbSet` it works fine. In fact reducing the time between `kdbGet` and `kdbSet` is good for avoiding conflicts anyway. There just can't be an unrelated `kdbGet` in between.

However, I agree, if there is a better solution, we should avoid additional restrictions.

Reply:
I agree it is the least-annoying restriction. But IIUC it always needs a ksDup on the user-side, so that a 3-way merge can be done if kdbSet fails.

@atmaxinger can you please give a small example how you imagined normal kdbSet usage (with conflict handling) to be with this restriction?

Reply:
> But IIUC kdbSet would always need a ksDup on the user-side, so that a 3-way merge can be done 

Where are you getting this idea from?
What does `ksDup` have to do with this alternative?
Why do you bring the 3-way merge into this?

Please answer this questions, I don't have any ideas on how to address this in the decision file.

> (with conflict handling)

What conflicts do you mean with that? 

> as it only decreases the likelyhood but doesn't give you any guarantee.

Aggree, but IMHO discussing merge conflict strategies in this decision is a bit off-topic.

Reply:
I have added a sample for kdbSet usage in 49c20d33

Reply:
> IIUC kdbSet would always need a ksDup on the user-side, so that a 3-way merge can be done

I'm not sure how the 3-way merge would be different for this option. From what I understand, you need at least **two** `ksDup` and possibly even a `ksDeepDup` to be on the safe side. To do a 3-way merge you need 3 keysets: base, ours and theirs. AFAIK "ours" is what we gave the `kdbSet` that failed. Then we do another `kdbGet` which gives us "theirs". And "base" is basically what this whole decision is about. For things to be correct, it must the data returned by the last `kdbGet` that **matched** the failed `kdbSet`.

> What conflicts do you mean with that?

I think in this case, it refers to the kinds of conflicts that make a `kdbSet` fail. The recommended way to resolve these is with a 3-way merge. Although I don't think we have a full example with explanations anywhere that definitely works in all cases.

The simplest case here is, if a file is edited on disk between `kdbGet` and `kdbSet`.

>  IIUC we normally have failures anymore in kdbSet but a successful kdbGet would tell us about conflicts...

No idea what this is supposed to mean...

Reply:
I am referring to examples/kdbset.c, what we currently recommend to do when someone wants to call `kdbSet`.

We are discussing about restrictions in sequences, so for me the question comes up if these examples still would work.

> What does ksDup have to do with this alternative?

See line 57 of examples/kdbset.c

https://github.com/ElektraInitiative/libelektra/blob/664bc1d21410afe92dfc3eb9b5921a017176e8fa/examples/kdbset.c#L57

with the current pattern (no kdbGet needed before), one could easily move the ksDup inside the `if` (if we use `if` instead of `while`, which we should probably recommend anyway).

Reply:
> No idea what this is supposed to mean...

A "won't" was missing in the sentence. I'll reexplain in more detail: Currently, earlier kdbSet from other processes will make kdbSet from our own process failing. If you call `kdbGet`, however, nothing fails and you simply pull in the changes from the other process.

Previous situation:

```
Process A                |           Process B
kdbSet() -> succeeds     |
                         |  kdbSet() -> fails with conflict
```

IIUC new situation:


```
Process A                |           Process B
kdbSet() -> succeeds     |
                         |  kdbGet() -> pulls in updates from Process A
                         |  kdbSet() -> succeeds
```

But there is no guarantee for that to happen, obviously also following could happen, even if kdbGet is directly before kdbSet:


```
Process A                |           Process B
                         |  kdbGet() -> NOP
kdbSet() -> succeeds     |
                         |  kdbSet() -> fails with conflict
```

So conflict handling would become more complicated, as there are two scenarios.

Reply:
> We are discussing about restrictions in sequences, so for me the question comes up if these examples still would work.

This option would put a restriction on the `parentKey`. That would not affect the 3-way merge example, because clearly you'll always use the same `parentKey` (at least the same name, if not pointer-equal).

> one could easily move the ksDup inside the if

Pretty sure the `ksDup` only exists to make cleanup easier.

> if we use if instead of while, which we should probably recommend anyway)

No, the `while` make sense. There could be repeated conflicts (e.g. if multiple process try to write at the same time).

Reply:
> So conflict handling would become more complicated, as there are two scenarios.

I don't see what changed. Everything you described should also apply equally to the current situation and to the new situation using option 2 or 3 from this decision. The only difference is in the "new situation" there is a new way for `kdbSet` to fail. When the keyset/parentKey doesn't match. But that is resolved just like the existing conflicts you described. Call `kdbGet` again (this time use the correct keyset) and try `kdbSet` once more.

See also https://github.com/ElektraInitiative/libelektra/pull/4574#issuecomment-1297391738

> It is wrong as it is, as the loop only makes sense if kdbSet failed because of an conflict, which isn't necessarily the case in this code.

But that doesn't actually make the `while` wrong. That only means there is a missing `if (hasConflict)` _around_ the `while`.



> But there are even two ksDup in this example:
> 
> https://github.com/ElektraInitiative/libelektra/blob/664bc1d21410afe92dfc3eb9b5921a017176e8fa/examples/kdbset.c#L63
> 
> And this second one is needed to have theirs and ours.

1. This comment chain is still on option 2, where the `parentKey` must match the previous `kdbSet`. Which makes the solution much simpler than for option 3. But again, see https://github.com/ElektraInitiative/libelektra/pull/4574#issuecomment-1297391738 for a solution that should also work for this option 2.
2. I'm 99% sure in the current `master` situation the example doesn't need this `ksDup` at all. At least not for 3-way merge reasons. It could just be `theirs = ksNew (0, KS_END)` if `kdbGet` would work as expected (It doesn't see [`doc/decisions/internal_cache.md`](https://github.com/ElektraInitiative/libelektra/blob/514319e26cfb62c3577df99d167985ecf9036d13/doc/decisions/internal_cache.md#fewer-keys)).

> We couldn't get it within the loop if we would have done the kdbGet already before kdbSet, at least in situation (1) of process examples above.

I think https://github.com/ElektraInitiative/libelektra/pull/4574#issuecomment-1297391738 should also cover this.

Reply:
Very messy discussion (also my fault), @atmaxinger please update the decision and let us go from there.

Reply:
A good example of why I suggested in https://github.com/ElektraInitiative/libelektra/pull/4562#discussion_r1010366283 that longer discussions should be moved to the root thread.

Reply:
Oh god ... where to even begin with summarizing this ...

Reply:
> Oh god ... where to even begin with summarizing this ...

This is why it is so important to constantly update and also lead discussions to directions that actually clarify questions relevant for you.